ambari-server/src/main/java/org/apache/ambari/server/controller/internal/ClusterTemplateArtifactPasswordReplacer.java

```diff
@@ -0,0 +1,138 @@
+    return passwordsReplaced;
+  }
+
+  /**
+   * Replaces passwords in the received cluster template artifact based on the received password information extracted
+   * from stacks.
+   * @param artifactData the cluster template artifact
+   * @param passwordProperties a multimap containing password type properties. The map has a structure of
+   *                           config type -> password properties.
+   * @return the cluster template artifact with passwords replaced
+   */
+  protected Map<String, Object> replacePasswordsInConfigurations(Map<String, Object> artifactData,
+                                                                 Multimap<String, String> passwordProperties) {
+    return (Map<String, Object>)applyToAllConfigurations(artifactData,
+      config -> {
+        Configuration configuration = Configurable.parseConfigs(config);
+        Configuration replacedConfiguration =
+          SecretReference.replacePasswordsInConfigurations(configuration, passwordProperties);
+        return Configurable.convertConfigToMap(replacedConfiguration);
+      }
+    );
+  }
+
+  /**
+   * <p> Recursively scans the received data structure consisting of maps, lists ans simple values (a parsed Json) and
+   * applies the @{code transform} function to each configurations found. </p>
+   * <p> A value counts as configuration if it has a type of @{link java.util.List} and is a value in a map with
+   * {@code "configurations"} key.
+   *
+   * </p>
+   * @param data the data to process recursively (is a structure of maps, lists and simple values)
+   * @param transform the transformation to apply to configuration values
+   * @return a replication of the input data with the transformation applied to all configuration values.
+   */
+  protected Object applyToAllConfigurations(Object data,
+                                            Function<List<Map<String, Object>>, Object> transform) {
+    // recursively call for lists
+    if (data instanceof List<?>) {
+      return ((List<Object>)data).stream().
+        map(item -> applyToAllConfigurations(item, transform)).collect(toList());
+    }
+    else if (data instanceof Map<?, ?>) {
+      return ((Map<String, Object>) data).entrySet().stream().map(
+        entry -> {
+          // apply transformation for configurations
+          if ("configurations".equals(entry.getKey()) && entry.getValue() instanceof List<?>) {
+            return new AbstractMap.SimpleEntry<>(
+              entry.getKey(),
+              transform.apply((List<Map<String, Object>>)entry.getValue()));
+          }
+          // recursively call for non-configuration elements
+          else {
+            return new AbstractMap.SimpleEntry<>(
+              entry.getKey(),
+              applyToAllConfigurations(entry.getValue(), transform));
+          }
+        }).collect(toMap(Map.Entry::getKey, Map.Entry::getValue));
```

Comment:
Can you please extract this lambda expression to a function that can be referenced and tested more easily?

Reply:
I am not sure it would improve readability. It is a simple if branch with one statement within each case; I think it is more self-explanatory to read in the context of the outer branch.

Reply:
I broke it up to smaller methods. I don't think it improves testing (the algorithm should be tested as a whole), on the other hand it hopefully improves readability.